consensus/safety-rules/src/safety_rules.rs

```diff
@@ -46,12 +49,75 @@ impl SafetyRules {
             .ok_or_else(|| Error::NotInitialized("validator_signer".into()))
     }
 
-    fn verifier(&self) -> Result<&ValidatorVerifier, Error> {
-        self.validator_verifier
+    fn epoch_state(&self) -> Result<&EpochState, Error> {
+        self.epoch_state
             .as_ref()
-            .ok_or_else(|| Error::NotInitialized("validator_verifier".into()))
+            .ok_or_else(|| Error::NotInitialized("epoch_state".into()))
     }
 
+    /// First voting rule
+    fn last_vote_round_check(&self, proposed_block: &BlockData) -> Result<(), Error> {
+        let last_voted_round = self.persistent_storage.last_voted_round()?;
+        if proposed_block.round() <= last_voted_round {
+            debug!(
+                "Vote proposal is old {} <= {}",
+                proposed_block.round(),
+                last_voted_round
+            );
+            return Err(Error::OldProposal {
+                proposal_round: proposed_block.round(),
+                last_voted_round: self.persistent_storage.last_voted_round()?,
+            });
+        }
+        Ok(())
+    }
+
+    /// Second voting rule
+    fn preferred_round_check(&mut self, quorum_cert: &QuorumCert) -> Result<(), Error> {
```

Comment:
verify_and_update or check_and_update

Reply:
should also update sign_timeout

Reply:
there's no qc in sign_timeout

Reply:
sorry comment was intended to go to the last voted round